File: mockets/test/cs/IntDataTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading;
using System.Net.Sockets;
using System.Net;
using System.IO;

using us.ihmc.mockets;

namespace IntDataTest
{
    class Util
    {
        public static long currentTimeMillis()
        {
            return DateTime.Now.Ticks / (long)10e3;
        }

        public static uint from4BytesToUnsignedInt(byte[] buffer, int offset)
        {
            uint result = 0;
            result = buffer[offset + 0];
            result = (result << 8) | buffer[offset + 1];
            result = (result << 8) | buffer[offset + 2];
            result = (result << 8) | buffer[offset + 3];

            return result;
        }

        public static void fromUnsignedIntTo4Bytes(uint value, byte[] buffer, int offset)
        {
            buffer[offset + 0]   = (byte) ((value >> 24) & 0xFF);
            buffer[offset + 1] = (byte) ((value >> 16) & 0xFF);
            buffer[offset + 2] = (byte) ((value >> 8) & 0xFF);
            buffer[offset + 3] = (byte) (value & 0xFF);
        }

        public static void saveStats(ManagedMessageMocket mocket, String type, long txTime)
        {
            String fname = "stats-" + type + "-mockets-cs.txt";
            FileStream file = new FileStream(fname, FileMode.OpenOrCreate, FileAccess.Write);
            StreamWriter sw = new StreamWriter(file);
            sw.WriteLine("[{0}]\t{1}\t{2}\t{3}\t{4}\t{5}\t{6}\t{7}\t{8}",
                Util.currentTimeMillis(),
                txTime,
                mocket.getStatistics().getSentPacketCount(),
                mocket.getStatistics().getSentPacketCount(),
                mocket.getStatistics().getSentByteCount(),
                mocket.getStatistics().getReceivedPacketCount(),
                mocket.getStatistics().getReceivedByteCount(),
                mocket.getStatistics().getRetransmitCount(),
                mocket.getStatistics().getDuplicatedDiscardedPacketCount() + mocket.getStatistics().getNoRoomDiscardedPacketCount()
                );

            sw.Flush();
            sw.Close();
            file.Close();
        }

        public static void saveStats(Socket socket, String type, long txTime)
        {
            String fname = "stats-" + type + "-sockets-cs.txt";
            FileStream file = new FileStream(fname, FileMode.OpenOrCreate, FileAccess.Write);
            StreamWriter sw = new StreamWriter(file);
            sw.WriteLine("[{0}]\t{1}",
                Util.currentTimeMillis(),
                txTime
                );

            sw.Flush();
            sw.Close();
            file.Close();
        }

        //public static void saveStats(Socket socket, String type, long txTime)
        public static void saveStats(TcpClient tcpClient, String type, long txTime)
        {
            String fname = "stats-" + type + "-sockets-cs.txt";
            FileStream file = new FileStream(fname, FileMode.OpenOrCreate, FileAccess.Write);
            StreamWriter sw = new StreamWriter(file);
            sw.WriteLine("[{0}]\t{1}",
                Util.currentTimeMillis(),
                txTime
                );

            sw.Flush();
            sw.Close();
            file.Close();
        }

        public static void reportError(String type, String errMSG)
        {
            String fname = "stats-" + type + "-mockets-cs.txt";
            FileStream file = new FileStream(fname, FileMode.OpenOrCreate, FileAccess.Write);
            StreamWriter sw = new StreamWriter(file);
            sw.WriteLine("[{0}]\t{1}",
                Util.currentTimeMillis(),
                errMSG
                );

            sw.Flush();
            sw.Close();
            file.Close();
        }
    } // class Util
    
    class Stats
    {
        public void update(double value)
        {
            _sumValues += value;
            _sumSqValues += (value * value);
            _totalNumValues++;
        }

        public void reset()
        {
            _sumValues = 0;
            _sumSqValues = 0;
            _totalNumValues = 0;
        }

        public int getNumValues()
        {
            return _totalNumValues;
        }

        public double getAverage()
        {
            return (_sumValues/_totalNumValues);
        }

        public double getStDev()
        {
            double avg = getAverage();
            double aux = (_totalNumValues * avg * avg)
                          - (2 * avg * _sumValues)
                          + _sumSqValues;
            aux = (double) aux / (_totalNumValues - 1);
            aux = Math.Sqrt (aux);
            return aux;
        }


        private double _sumValues   = 0.0;
        private double _sumSqValues = 0.0;
        private int _totalNumValues = 0;
    } //class Stats

    
    class ClientConnHandler
    {
        public ClientConnHandler (ManagedMessageMocket msgMocket)
        {
            _mocket = msgMocket;
            PeerUnreachableWarningCallback callback = new PeerUnreachableWarningCallback(this.peerUnreachableCallback);
            _mocket.registerPeerUnreachableWarningCallback(callback);
            _useMockets = true;
        }
/*
        public ClientConnHandler (Socket socket)
        {
            _socket = socket;
            _useMockets = false;
        }
*/
        public ClientConnHandler(TcpClient tcpClient)
        {
            _tcpClient = tcpClient;
            _useMockets = false;
        }

        public void run()
        {
            Console.WriteLine("ConnectionHandler: client handler thread started");
            long startTime = Util.currentTimeMillis();
            byte[] buffer = new byte[IntDataTest.BUFFER_SIZE];
            byte[] chArray = new byte[] {(byte)'.'};
            uint totalToRead = 0;
            uint totalRead = 0;

            if (_useMockets) {
                _mocket.receive(buffer, 4, 0);

                totalToRead = Util.from4BytesToUnsignedInt(buffer, 0);
                Console.WriteLine("IntDataTest:Server >> will read a total of [" + totalToRead + "] bytes. Remote port  = " + _mocket.getRemotePort());                
                while ((totalRead < totalToRead) && !_closeConn) {
                    //System.out.println(" will try to read.");
                    uint numRead = (uint)_mocket.receive (buffer, (uint)buffer.Length);
                    if (numRead < 0) {
                        Console.WriteLine("error reading????");
                    }
                    totalRead += numRead;
                    //Console.WriteLine("read so far:: {0}", totalRead);
                }
                
                Console.WriteLine("received " + totalRead + " bytes.");
            }
            else {
                NetworkStream stream = _tcpClient.GetStream();
                Console.WriteLine ("before socket receive....");
                //_socket.Receive (buffer);
                stream.Read(buffer, 0, 4);
                Console.WriteLine ("after socket receive...");
                totalToRead = Util.from4BytesToUnsignedInt (buffer, 0);
                Console.WriteLine("IntDataTest:ServerSocket >> will read a total of [" + totalToRead + "] bytes. ");
                while ((totalRead < totalToRead)) {
                    //System.out.println(" will try to read.");
                    //uint numRead = (uint) _socket.Receive (buffer);
                    uint numRead = (uint)stream.Read(buffer, 0, buffer.Length);
                    if (numRead < 0) {
                        Console.WriteLine ("error reading????");
                    }
                    //Console.WriteLine("just read:: {0}", numRead);
                    totalRead += numRead;
                    //Console.WriteLine("socket read so far:: {0}", totalRead);
                }
            }

            long totalTime = Util.currentTimeMillis() - startTime;
            Console.WriteLine("Socket Done reading");

            if (_closeConn) {
                Console.WriteLine("Connection reset.");
                Util.reportError ("server", "Connection Reset. 1");
                return;
            }
            if (_useMockets) {
                ManagedMessageSender sender = _mocket.getSender (true, true);
                sender.send (chArray, (uint)chArray.Length);
                Console.WriteLine ("IntDataTest: Sending ACK!");
                _mocket.close();
                Util.saveStats (_mocket, "MessageServer", totalTime);
            }
            else {
                Console.WriteLine ("IntDataTest :: sending back the ACK [" + totalRead + "]");
                //_socket.Send (chArray);                
                NetworkStream stream = _tcpClient.GetStream();
                stream.Write(chArray, 0, 1);
                stream.Flush();

                Console.WriteLine("IntDataTest: Socket Sending ACK!");
                //_socket.Close();
                _tcpClient.Close();
                Util.saveStats (_socket, "SocketServer", totalTime);
            }
        }

        public void Start()
        {
            Thread t = new Thread(new ThreadStart(run));
            t.Start();
        }

        public Boolean peerUnreachableCallback(uint timeSinceLastContact)
        {
            Console.WriteLine("\nIntDataTest :: PeerUnreachableWarning called :: {0}", timeSinceLastContact);
            if (timeSinceLastContact > 15000) 
            {
                _closeConn = true;
                return true;
            }
            return false;
        }

        private Boolean _useMockets;
        private ManagedMessageMocket _mocket;
        private Socket _socket;
        private TcpClient _tcpClient;
        private bool _closeConn = false;
    }

    class IntDataTest
    {
        public void runServerMocket()
        {
            ManagedMessageServerMocket serverMocket = new ManagedMessageServerMocket();
            serverMocket.listen(_portNum);

            while (true) {
                Console.WriteLine("serverMocket:: waiting for connections...");
                ManagedMessageMocket mocket = serverMocket.accept();
                Console.WriteLine("serverMocket:: connection Accepted.");
                ClientConnHandler connHandler = new ClientConnHandler(mocket);
                connHandler.Start();
            }
        }

        public void runServerSocket()
        {
            IPAddress bindIP = IPAddress.Any;
            TcpListener serverSocket = new TcpListener(bindIP, _portNum);
            serverSocket.Start();
            while (true) {
                //Socket socket = serverSocket.AcceptSocket();
                //ClientConnHandler connHandler = new ClientConnHandler(socket);
                TcpClient client = serverSocket.AcceptTcpClient();
                ClientConnHandler connHandler = new ClientConnHandler(client);

                connHandler.Start();
            }
        }

        public void runServers(ushort portNumber)
        {
            _portNum = portNumber;
            
            Thread serverMocketThread = new Thread(new ThreadStart(runServerMocket));
            serverMocketThread.Name = "ServerMocketThread";

            Thread serverSocketThread = new Thread(new ThreadStart(runServerSocket));
            serverSocketThread.Name = "ServerSocketThread";

            serverMocketThread.Start();
            serverSocketThread.Start();
        }

        public Boolean peerUnreachableCallback(uint timeSinceLastContact)
        {
            Console.WriteLine("\nIntDataTest :: PeerUnreachableWarning(2) called :: {0}", timeSinceLastContact);
            return (timeSinceLastContact > 15000);
        }

        public int doClientTask (string remoteHost, ushort remotePort, bool useMockets, Stats stats)
        {
            int rc;
            byte[] buf = new byte[1024];

            if (useMockets) {
                Console.WriteLine ("doClientTask: Using MessageMockets");
                ManagedMessageMocket mocket = new ManagedMessageMocket();
                //mocket.bind("127.0.0.1", 6789);

                Console.WriteLine ("doClientTask: MessageMockets: Before connect\n");
                rc = mocket.connect(remoteHost, remotePort);
                if (0 != rc) {
                    Console.WriteLine("doClientTask: failed to connect using MessageMockets to remote host {0} on port {1}; rc = {2}",
                             remoteHost, remotePort, rc);
                    Console.WriteLine("doClientTask: Unable to connect\n");
                    return -1;
                }

                PeerUnreachableWarningCallback callback = new PeerUnreachableWarningCallback(this.peerUnreachableCallback);
                mocket.registerPeerUnreachableWarningCallback(callback);

                int dataSize = IntDataTest.DATA_SIZE;
                int bytesSent = 0;
                long startTime = Util.currentTimeMillis();

                ManagedMessageSender sender = mocket.getSender(true, true);
                byte[] auxBuf = new byte[4];
                Util.fromUnsignedIntTo4Bytes((uint)dataSize, auxBuf, 0);

                // sending the data size.
                sender.send(auxBuf, 4);
                Console.WriteLine("will write {0} bytes of data", dataSize);
                while (bytesSent < dataSize) {
                    sender.send (buf, (uint)buf.Length);
                    bytesSent += buf.Length;
                    //Console.WriteLine("written so far :: {0}", bytesSent);
                }

                Console.WriteLine("done writing data.");

                mocket.receive(auxBuf, 1);
                if (auxBuf[0] != '.') {
                    Console.WriteLine("doClientTask: failed to receive . from remote host");
                    return -2;
                }

                long time = Util.currentTimeMillis() - startTime;
                stats.update((double) time);

                // Save results to a file
                Util.saveStats (mocket, "MsgMockets-cs", time);
            }
            else {
                //Socket socket = new Socket (AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                //socket.Connect(remoteHost, remotePort);
                TcpClient client = new TcpClient(remoteHost, remotePort);
                NetworkStream stream = client.GetStream();

                int dataSize = IntDataTest.DATA_SIZE;
                int bytesSent = 0;
                long startTime = Util.currentTimeMillis();

                byte[] auxBuf = new byte[4];
                Util.fromUnsignedIntTo4Bytes((uint)dataSize, auxBuf, 0);

                // sending the data size.
                //socket.Send (auxBuf);
                stream.Write(auxBuf, 0, 4);
                Console.WriteLine ("socket will write {0} bytes of data", dataSize);
                while (bytesSent < dataSize) {
                    //socket.Send (buf);
                    stream.Write(buf, 0, buf.Length);
                    bytesSent += buf.Length;
                    //Console.WriteLine("written so far :: {0}", bytesSent);
                }
                stream.Flush();

                Console.WriteLine("Socket done writing data.");
                Console.WriteLine("->>:Socket before socket.Receive");
                //socket.Receive (auxBuf);
                stream.Read(auxBuf, 0, 1);
                Console.WriteLine("->>:Socket after socket.Receive", auxBuf[0]);
                if (auxBuf[0] != '.') {
                    Console.WriteLine("doClientTask:Socket failed to receive . from remote host");
                    return -2;
                }

                long time = Util.currentTimeMillis() - startTime;
                stats.update ((double)time);

                // Save results to a file
                Console.WriteLine("->>:Socket before saveStats");
                //Util.saveStats (socket, "Sockets-cs", time);
                Util.saveStats(client, "Sockets-cs", time);
                Console.WriteLine("->>:Socket after saveStats");
            }
            return 0;
        } // doClientTask

        // ------------------------------------------------------------------------------
        private ushort _portNum = 4000;
        public const int BUFFER_SIZE = 1024;
        public const int DATA_SIZE = 1024 * 1024;
        // ------------------------------------------------------------------------------

        static void Main(string[] args)
        {
            if (args.Length < 2) {
                Console.WriteLine("usage: <server|client> port [<remoteHost> [iterations]]");
                System.Environment.Exit(1);
            }

            int portNumber = Int32.Parse(args[1]);
            string remoteHost = "localhost";
            int iterations = 100;

            IntDataTest idt = new IntDataTest();

            if (args.Length > 2) {
                remoteHost = args[2];
            }
            if (args.Length > 3) {
                iterations = Int32.Parse(args[3]);
            }

            if (args[0] == "client") {
                Console.WriteLine("Trying to establish a mocket/socket connection to port {0}", portNumber);
                Stats socketStats = new Stats();
                Stats mocketStats = new Stats();
                int rc;

                for (int i = 0; i < iterations; i++)
                {
                    Console.WriteLine("iteration # :: {0}", i);

                    rc = idt.doClientTask(remoteHost, (ushort) portNumber, true, mocketStats);
                    if (rc != 0)
                    {
                        Console.WriteLine("main: doClientTask failed for sockets with rc = {0}", rc);
                        System.Environment.Exit(-2);
                    }
                    
                    rc = idt.doClientTask(remoteHost, (ushort) portNumber, false, socketStats);
                    if (rc != 0)
                    {
                        Console.WriteLine("main: doClientTask failed for mockets with rc = {0}", rc);
                        System.Environment.Exit(-3);
                    }

                    Console.WriteLine("-----------------------------------------");
                    Console.WriteLine("TotalAttempts: {0}\n", i + 1);
                    Console.WriteLine("MessageMocket Stats:: Average:       {0,10:#.##}", mocketStats.getAverage());
                    Console.WriteLine("MessageMocket Stats:: St Deviation:  {0,10:#.##}", mocketStats.getStDev());
                    Console.WriteLine("Socket Stats:: Average:       {0,10:#.##}", socketStats.getAverage());
                    Console.WriteLine("Socket Stats:: St Deviation:  {0,10:#.##}", socketStats.getStDev());
                    //Console.WriteLine("Socket Stats:: St Deviation:  %10.4f", socketStats.getStDev());
                    Console.WriteLine("-----------------------------------------");

                    Console.WriteLine("Sleeping for 3 seconds...\n");
                    Thread.Sleep(3000);
                }
            }
            else if (args[0] == "server") {
                idt.runServers((ushort)portNumber);
            }
        } //Main
    }
}

File: mockets/test/java/IntDataTest.java
import us.ihmc.mockets.MessageServerMocket;
import us.ihmc.mockets.MessageMocket;
import us.ihmc.mockets.Mocket;
import us.ihmc.mockets.MocketStatusListener;
import us.ihmc.mockets.MessageServerMocket;
import us.ihmc.mockets.Sender;

import us.ihmc.util.ByteConverter;

import java.io.BufferedWriter;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.net.ServerSocket;
import java.util.Random;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author: Maggie Breedy
 * @version $Revision: 1.4 $
 *
 */
public class IntDataTest
{
    public static class ServerMessageMocketThread extends Thread
    {
        public ServerMessageMocketThread (MessageServerMocket msm)
        {
            _msgServerMocket = msm;
        }

        public void run()
        {
            System.out.println ("ServerMocketThread ::run");
            while (true) {
                try {
                    System.out.println ("ServerMocketThread ::before accept");
                    MessageMocket msgMocket = _msgServerMocket.accept();
                    System.out.println ("ServerMocketThread ::after accept");
                    System.out.println ("ServerMocketThread :: accepted a Message Mocket connection.");
                    new ConnHandler (msgMocket).start();
                }
                catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }

        MessageServerMocket _msgServerMocket;
    } //ServerMocketThread

    /**
     *
     */
    public static class ServerSocketThread extends Thread
    {
        public ServerSocketThread(ServerSocket ss)
        {
            _serverSocket = ss;
        }

        public void run()
        {
            while (true) {
                try {
                    Socket socket = _serverSocket.accept();
                    System.out.println ("ServerSocketThread :: accepted a Socket connection.");
                    new ConnHandler (socket).start();
                }
                catch (Exception ex) {
                    ex.printStackTrace();
                }
            }
        }

        ServerSocket _serverSocket;
    } //ServerSocketThread

    /**
     *
     */
    public static class ConnHandler extends Thread implements MocketStatusListener
    {
        public ConnHandler (MessageMocket msgMocket)
        {
            _msgMocket = msgMocket;
            _useMsgMockets = true;
        }

        public ConnHandler(Socket socket)
        {
            _socket = socket;
            _useMsgMockets = false;
        }

        public boolean peerUnreachableWarning(long x)
        {
            _closeConn = (x > 60 * 1000);

            if (_closeConn) {
                System.out.println("AdHocTest :: MSL >> connection inactive for 60 seconds, will disconnect.");
            }

            return _closeConn;
        }

        public void run()
        {
            try {
                InetSocketAddress remoteAddr;

                if (_useMsgMockets) {
                    remoteAddr = (InetSocketAddress) _msgMocket.getPeerName();
                    //_msgMocket.addMocketStatusListener(this);
                }
                else {
                    remoteAddr = (InetSocketAddress) _socket.getRemoteSocketAddress();
                }

                System.out.println ("AdHocTest :: accepted a connection from " + remoteAddr.getAddress().getHostAddress()
                                    + ":" + remoteAddr.getPort());

                long startTime = System.currentTimeMillis();
                byte[] buffer = new byte[BUFFER_SIZE];
                int numBytes = 0;
                int numRead = 0;
                int totalRead = 0;
                OutputStream os = null;
                _closeConn = false;

                byte[] chArray = new byte [1];
                chArray[0] = '.';

                if (_useMsgMockets) {
                    //numRead = _msgMocket.receive (buffer, 0, buffer.length);
                    _msgMocket.receive (buffer, 0, 4);
                    int totalToRead = (int) ByteConverter.from4BytesToUnsignedInt (buffer, 0);

                    System.out.println ("AdHocTest :: Server >> will read a total of [" + totalToRead + "] bytes. ");
                    while ((totalRead < totalToRead) && !_closeConn) {
                        //System.out.println(" will try to read.");
                        numRead = _msgMocket.receive (buffer, 0, buffer.length);
                        if (numRead < 0) {
                            System.out.println ("error reading????");
                        }
                        totalRead += numRead;
                        //System.out.println("total read ::: " + totalRead);
                    }
                    System.out.println("received " + totalRead + ". Closing the connection.");
                }
                else {
                    InputStream is = _socket.getInputStream();
                    os = _socket.getOutputStream();

                    DataInputStream dis = new DataInputStream(is);
                    numBytes = dis.readInt();
                    System.out.println ("AdHocTest :: Server >> will read a total of [" + numBytes + "] bytes. ");
                    while ((totalRead < numBytes) && !_closeConn) {
                        try {
                            numRead = is.read (buffer);
                            if (numRead > 0) {
                                totalRead += numRead;
                            }
                            else {
                                System.out.println("received " + numRead + ". Closing the connection.");
                                _closeConn = true;
                            }
                        }
                        catch (Exception ex) {
                            ex.printStackTrace();
                            reportError ("server", "Connection Reset (failed to write.)");
                            break;
                        }
                    }
                }

                long totalTime = System.currentTimeMillis() - startTime;
                System.out.println ("AdHocTest :: done reading.");

                if (_closeConn) {
                    System.out.println ("Connection reset.");
                    reportError ("server", "Connection Reset. 1");
                    return;
                }
                if (_useMsgMockets) {
                    Sender sender = _msgMocket.getSender (true, true);
                    sender.send (chArray, 0, chArray.length);
                    //System.out.println ("--->>Sending ACK!");
                    _msgMocket.close();
                    saveStats (_msgMocket, "MessageServer", totalTime);
                }
                else {
                    System.out.println ("AdHocTest :: sending back the ACK [" + totalRead + "]");
                    DataOutputStream dos = new DataOutputStream(os);
                    dos.writeInt (totalRead);
                    _socket.close();
                    saveStats(_socket, "server", totalTime);
                }
            }
            catch (Exception ex) {
                ex.printStackTrace();
            }

        }

        MessageMocket _msgMocket;
        Socket _socket;
        boolean _useMsgMockets = true;
        boolean _closeConn;
    } //class ConnHandler

    /**
     *
     */
    public static class ClientThread extends Thread implements MocketStatusListener
    {
        public ClientThread (String remoteHost, int port, boolean useMockets, Stats stats)
        {
            _remoteHost = remoteHost;
            _port = port;
            _useMsgMockets = useMockets;
            _stats = stats;
        }

        public boolean peerUnreachableWarning (long x)
        {
            _closeConn = (x > 40 * 1000);

            if (_closeConn) {
                System.out.println ("AdHocTest :: MSL >> connection inactive for 60 seconds, will disconnect.");
            }

            return _closeConn;
        }

        public void run()
        {
            String remoteHost = _remoteHost;
            int portNumber = _port;

            try {
                InetAddress remoteAddr = InetAddress.getByName (remoteHost);
                System.out.println("AdHocTest :: will try to connect to " + remoteHost + ":" + portNumber);

                MessageMocket _msgMocket = null;
                Socket socket = null;
                OutputStream os = null;;
                InputStream is = null;

                byte[] buffer = new byte [BUFFER_SIZE];
                Random rand = new Random (1234);
                for (int i = 0; i < buffer.length; i++) {
                    buffer[i] = (byte) rand.nextInt();
                }

                Sender sender = null;
                int numWritten = 0;
                long startTime = 0;
                long totalTime = 0;

                if (_useMsgMockets) {
                    System.out.println ("USING MESSAGE MOCKETS.");
                    _msgMocket = new MessageMocket();
                    //_msgMocket.connect (remoteAddr, portNumber, _connectTimeout);
                    InetSocketAddress isa = new InetSocketAddress(remoteAddr, portNumber);
                    System.out.println(" trying to connect to.... " + isa.toString());
                    _msgMocket.connect(isa);

                    //_msgMocket.addMocketStatusListener (this);
                    System.out.println("AdHocTest :: connected successfully " + remoteHost + ":" + portNumber);
                    System.out.println ("-------USING MESSAGE MOCKETS: " + _msgMocket);
                    sender = _msgMocket.getSender (true, true);
                    startTime = System.currentTimeMillis();
                    _closeConn = false;

                    byte[] buffAux = new byte[4];
                    ByteConverter.fromUnsignedIntTo4Bytes (_numBytesToSend, buffAux, 0);
                    sender.send (buffAux, 0, buffAux.length);

                    System.out.println("AdHocTest :: will start sending data.");
                    while ((numWritten < _numBytesToSend) && !_closeConn) {
                         sender.send (buffer, 0, buffer.length);
                         numWritten += buffer.length;
                         //System.out.println ("numWritten >>>>>> " + numWritten);
                    }
                }
                else {
                    System.out.println("USING SOCKETS.");
                    SocketAddress sockAddr = new InetSocketAddress (InetAddress.getByName(remoteHost),
                                                                    portNumber);
                    socket = new Socket();
                    socket.connect(sockAddr, _connectTimeout);
                    System.out.println("AdHocTest :: connected successfully " + remoteHost + ":" + portNumber);

                    os = socket.getOutputStream();
                    is = socket.getInputStream();
                    DataOutputStream dos = new DataOutputStream(os);
                    dos.writeInt (_numBytesToSend);

                    startTime = System.currentTimeMillis();
                    _closeConn = false;
                    System.out.println("AdHocTest :: will start sending data.");
                    while ((numWritten < _numBytesToSend) && !_closeConn) {
                        try {
                            os.write(buffer);
                            numWritten += buffer.length;
                        }
                        catch (Exception ex) {
                            ex.printStackTrace();
                            reportError("socket-client", "Connection Reset (failed to write.)");
                            break;
                        }
                    }
                }

                if (_closeConn) {
                    System.out.println("connection reset");
                    if (_useMsgMockets) {
                        reportError("msgMocket-client", "Connection Reset");
                    }
                    else {
                        reportError("socket-client", "Connection Reset");
                    }
                }
                else {
                    if (_useMsgMockets) {
                        byte[] chReply = new byte [1];
                        _msgMocket.receive (chReply, 0, chReply.length);
                        if (chReply[0] != '.') {
                            System.out.println ("doClientTask: failed to receive . from remote host\n");
                            return;
                        }
                        totalTime = System.currentTimeMillis() - startTime;
                        System.out.println ("->done sending data... it took [" + totalTime + "].");
                        saveStats (_msgMocket, "MsgMocket-client", totalTime);
                        _stats.update (totalTime);
                        _msgMocket.close();
                    }
                    else {
                        DataInputStream dis = new DataInputStream(is);
                        int totalRead = dis.readInt();
                        System.out.println ("AdHocTest :: server received a total of [" + totalRead + "] bytes.");
                        totalTime = System.currentTimeMillis() - startTime;
                        System.out.println ("done sending data... it took [" + totalTime + "]. Waiting for ACK from server.");
                        saveStats (socket, "socket-client", totalTime);
                        _stats.update (totalTime);
                        socket.close();
                    }
                }
            }
            catch (Exception ex) {
                ex.printStackTrace();
                if (_useMsgMockets) {
                    reportError ("msgMocket-client", "Connection Failed");
                }
                else {
                    reportError ("socket-client", "Connection Failed");
                }
            }

            System.out.println ("\n-----\n");
        }

        String _remoteHost;
        int _port;
        boolean _useMsgMockets;
        public boolean _closeConn;
        Stats _stats;
    } // ClientThread

    /**
     *
     */
    public synchronized static void saveStats (MessageMocket msgMocket, String type, long txTime)
    {
        String outFile = "stats-" + type + "-java.txt";

        try {
            FileOutputStream fos = new FileOutputStream (outFile, true);
            BufferedWriter bw = new BufferedWriter (new OutputStreamWriter(fos));

            Mocket.Statistics stats = msgMocket.getStatistics();
            String line = "[" + System.currentTimeMillis() + "]\t"
                        + "\t" + txTime
                        + "\t" + stats.getSentPacketCount()
                        + "\t" + stats.getSentByteCount()
                        + "\t" + stats.getReceivedPacketCount()
                        + "\t" + stats.getReceivedByteCount();
                       // + "\t" + stats.getRetransmittedPacketCount()
                       // + "\t" + stats.getDiscardedPacketCount();

            bw.write(line);
            bw.newLine();
            bw.flush();

            bw.close();
            fos.close();
        }
        catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    /**
     *
     */
    public synchronized static void saveStats (Socket socket, String type, long txTime)
    {
        String outFile = "stats-" + type + "-java.txt";

        try {
            FileOutputStream fos = new FileOutputStream (outFile, true);
            BufferedWriter bw = new BufferedWriter (new OutputStreamWriter(fos));

            String line = "[" + System.currentTimeMillis() + "]\t" + "\t" + txTime;

            bw.write (line);
            bw.newLine();
            bw.flush();

            bw.close();
            fos.close();
        }
        catch (Exception ex) {
            ex.printStackTrace();
        }
    }


    /**
     *
     */
    public synchronized static void reportError (String type, String errMSG)
    {
        String outFile = "stats-" + type + "-java.txt";

        try {
            FileOutputStream fos = new FileOutputStream (outFile, true);
            BufferedWriter bw = new BufferedWriter (new OutputStreamWriter(fos));

            String line = "[" + System.currentTimeMillis() + "]\t"
                        + "\t[" + errMSG + "]";
                        ;

            bw.write(line);
            bw.newLine();
            bw.flush();

            bw.close();
            fos.close();
        }
        catch (Exception ex) {
            ex.printStackTrace();
        }
    }
    // /////////////////////////////////////////////////////////////////////////
    // MAIN METHOD /////////////////////////////////////////////////////////////
    // /////////////////////////////////////////////////////////////////////////

    /**
     *
     */
    public static void main (String[] args) throws Exception
    {
        String usageMsg = "IntDataTest <client|server> <portnumber> [remotehost [iterations]]";

        if (args.length < 2) {
            System.out.println(usageMsg);
            System.exit(1);
        }

        int portNumber = Integer.parseInt(args[1]);
        int iterations = 100;
        String remoteHost = "localhost";

        if (args.length > 2) {
            remoteHost = args[2];
        }
        if (args.length > 3) {
            iterations = Integer.parseInt(args[3]);
        }
        Logger logger = Logger.getLogger ("us.ihmc.mockets");
        logger.setLevel (Level.OFF);

        if (args[0].equals ("server")) {
            MessageServerMocket msm = new MessageServerMocket (portNumber);
            System.out.println ("AdHocTest :: created a MessageServerMocket on port " + portNumber);
            ServerSocket ss = new ServerSocket ( portNumber);
            System.out.println ("AdHocTest :: created a ServerSocket on port " + portNumber);

            ServerMessageMocketThread smt = new ServerMessageMocketThread (msm);
            ServerSocketThread sst = new ServerSocketThread (ss);

            smt.start();
            sst.start();

            smt.join();
            sst.join();
        }
        else if (args[0].equals ("client")) {
            Stats mocketStats = new Stats();
            Stats socketStats = new Stats();

            for (int i = 0; i < iterations; i++) {
                ClientThread ct = new ClientThread (remoteHost, portNumber, true, mocketStats);
                ct.run();

                ct = new ClientThread (remoteHost, portNumber, false, socketStats);
                ct.run();
//                ct.start();
//                ct.join(60*1000); // wait at most 60 seconds for the thread to die.

                System.out.println ("-----------------------------------------");
                System.out.println ("TotalAttempts: " + (i + 1));
                System.out.println ("Mocket Stats:: Average:       " + mocketStats.getAverage());
                System.out.println ("Mocket Stats:: St Deviation:  " + mocketStats.getStDev());
                System.out.println ("Socket Stats:: Average:       " + socketStats.getAverage());
                System.out.println ("Socket Stats:: St Deviation:  " + socketStats.getStDev());
                System.out.println ("-----------------------------------------");
                System.out.println ("Sleeping for 10 seconds.... ");
                Thread.sleep (10000);
            }

            System.out.println ("\nFinished the test!!!.\n");
            System.exit(1);
        }
        else {
            System.out.println (usageMsg);
            System.exit (1);
        }
    } //main.

    // /////////////////////////////////////////////////////////////////////////
    // INTERNAL CLASSES ////////////////////////////////////////////////////////
    // /////////////////////////////////////////////////////////////////////////
    /**
     *
     */
    public static class Stats
    {
        public void update(double value)
        {
            _sumValues += value;
            _sumSqValues += Math.pow(value, 2);
            _totalNumValues++;
        }

        public void reset()
        {
            _sumValues = 0;
            _sumSqValues = 0;
            _totalNumValues = 0;
        }

        public int getNumValues()
        {
            return _totalNumValues;
        }

        public double getAverage()
        {
            return (_sumValues/_totalNumValues);
        }

        public double getStDev()
        {
            double avg = getAverage();
            double aux = (_totalNumValues*avg*avg)
                         - (2 * avg * _sumValues)
                         + _sumSqValues
                         ;
            aux = (double)aux / (_totalNumValues - 1);
            aux = Math.sqrt(aux);
            return aux;
        }

        double _sumValues = 0;
        double _sumSqValues = 0;
        int _totalNumValues = 0;
    } //class Stats

    // /////////////////////////////////////////////////////////////////////////

    private static int _connectTimeout = 15 * 1000;
    private static int _numBytesToSend = 1024*1024; // 1 Mb

    private static final int BUFFER_SIZE = 1024;
}
